crates/exex/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+use futures_util::{future::BoxFuture, FutureExt};
+use reth_db::database::Database;
+use reth_primitives::BlockNumber;
+use reth_provider::{
+    AccountReader, BlockReaderIdExt, CanonStateNotification, CanonStateNotifications,
+    CanonStateSubscriptions, ChainSpecProvider, ChangeSetReader, DatabaseProviderFactory,
+    EvmEnvProvider, StateProviderFactory,
+};
+use std::future::Future;
+use tokio::sync::watch;
+
+/// The handle to the ExEx.
+///
+/// The spawner of the ExEx is responsible for monitoring the ExEx and reacting to the changes in
+/// the height of the last processed block.
+#[derive(Debug, Clone)]
+pub struct ExExHandle {
+    /// Meta information about the ExEx.
+    pub meta: ExExMeta,
+    /// A watch channel that will receive the height of the last block processed by the ExEx.
+    ///
+    /// ExEx should guarantee that it will not require all earlier blocks in the future, meaning
+    /// that Reth is allowed to prune them.
+    ///
+    /// On reorgs, it's possible for the height to go down.
+    pub finished_height: watch::Receiver<BlockNumber>,
+}
+
+/// The main trait for the custom ExEx.
+///
+/// The ExEx should listen to the canonical state notifications and process them, updating the
+/// `finished_height` in [ExExHandle] accordingly.
+pub trait IntoExEx<DB: Database, P: FullProvider<DB>> {
+    /// Spawns the ExEx.
+    ///
+    /// The ExEx should listen to the canonical state notifications and process them, updating the
+    /// `finished_height` in [ExExHandle] accordingly. The ExEx may receive the same notification
+    /// again, and ExEx's reaction to that should be idempotent. Example cases when a duplicate
+    /// notification is possible are:
+    /// 1. Reth sends a notification during the pipeline sync, but restarts before committing the
+    /// data to the database.
+    /// 2. The ExEx doesn't update the `finished_height` in the handle.
+    ///
+    /// Returns a future that should be polled to advance the ExEx and a handle to the ExEx.
+    fn spawn_exex(
+        self,
+        provider: P,
+        canon_state_notifications: CanonStateNotifications,
+    ) -> (BoxFuture<'static, ()>, ExExHandle);
+}
+
+/// The meta information about the ExEx.
+#[derive(Debug, Clone)]
+pub struct ExExMeta {
+    /// The identifier of the ExEx. It will appear in logs and metrics.
+    pub id: String,
+    /// The height from which the ExEx should start processing blocks. Earlier blocks will not be
+    /// sent to the ExEx.
+    ///
+    /// In case when the ExEx needs all blocks from the genesis, it should be set to `0`.
+    pub from_height: BlockNumber,
+}
+
+/// The managed version of the [IntoExEx] trait that covers the common use case and simplifies the
+/// usage.
+pub trait ExEx<DB: Database, P: FullProvider<DB>>: Send + Sync + Unpin + 'static {
```

Comment:
`IntoExEx` is the stream based API

Reply:
the idea is to have a method on the builder like `add_exex(exex: impl IntoExEx)`, so you can pass either simplified `ExEx` or the actual `IntoExEx` impl.

agree that we should move this to the builder care as it reuses its components such as context and provider

Reply:
> perhaps ExEx should be a stream instead that returns updates

I'm not sure if we want this, but I can see that ExEx may want to communicate with Reth in other ways than just a `finished_height`, so I think making it outputting the stream of events with a variant `FinishedHeight(BlockNumber)` is a good idea.

Reply:
i don't see how else it would need to communicate, so imo we should leave it out unless there is a need

all we need to know from the exex is what we can prune or not, beyond that we don't really know what it's doing

Reply:
then this should be over the existing traits we have in builder, they are designed for this, see existing hook traits, the exex trait itself shouldn't require any generics if we just make it future/stream